src/Features/Core/Diagnostics/BaseDiagnosticIncrementalAnalyzer.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal abstract class BaseDiagnosticIncrementalAnalyzer : IIncrementalAnalyzer
```

Comment:
So does this suggest you expect us to implement a completely new DiagnosticIncrementalAnalyzer for the new approach? The work that John and I plan to do mainly addresses how to tweak DiagnosticAnalyzerDriver type to use compiler event queue for executing actions.
My concern with completely forking the DiagnosticIncrementalAnalyzer for different engines is that any fixes that goes into current one has to be ported to new one manually, and especially any big refactoring or bug fix you are doing in the incremental analyzer is going to double our work.
I was hoping that to start with we will just have a DiagnosticAnalyzerDriver2 or something such and a switch to turn on that driver. Once that work is done, fixing the various other IDE issues (compilation end action diagnostics, no duplicate callbacks, etc.) can be addressed as bug fixes.